File: elixir/lib/y2025/day06.ex
# ================ Trash Compactor ================

defmodule Y2025.Day06 do
  def part1(_ctx, lines) do
    do_part(lines, &parse_problems/1)
  end

  def part2(_ctx, lines) do
    do_part(lines, &parse_problems_the_cephalopod_way/1)
  end

  defp do_part(lines, parse_func) do
    parse_func.(lines)
    |> Enum.map(fn {numbers, operation} -> Enum.reduce(numbers, operation) end)
    |> Enum.sum()
  end

  # ================ helpers ================

  @doc """
  Returns a list of {numbers, operation} tuples.
  """
  def parse_problems(lines), do: parse_problems(lines, [])

  # Returns a list of {numbers, operation} tuples.
  defp parse_problems([], problems), do: problems

  defp parse_problems([line | lines], problems) do
    vals = String.split(line, " ", trim: true)

    new_problems =
      if Enum.at(vals, 0) in ["+", "*"] do
        add_operations(vals, problems)
      else
        add_numbers(vals, problems)
      end

    parse_problems(lines, new_problems)
  end

  # Add operations to problems.
  defp add_operations(vals, problems) do
    operations = vals |> Enum.map(fn s -> if(s == "+", do: &+/2, else: &*/2) end)

    problems
    |> Enum.zip(operations)
    |> Enum.map(fn {{numbers, _}, op} -> {numbers, op} end)
  end

  # Add numbers, but we don't have any problems yet so create empty ones.
  defp add_numbers(vals, []) do
    add_numbers(vals, Enum.map(vals, fn _ -> {[], nil} end))
  end

  # Add numbers to problems.
  defp add_numbers(vals, problems) do
    numbers = vals |> Enum.map(&String.to_integer/1)

    problems
    |> Enum.zip(numbers)
    |> Enum.map(fn {{numbers, op}, num} -> {[num | numbers], op} end)
  end

  @doc """
  Returns a list of {numbers, operation} tuples.
  """
  def parse_problems_the_cephalopod_way(_lines) do
    # TODO
  end
end

File: ruby/y2025/day06.rb
#!/usr/bin/env ruby
#
# Trash Compactor

require_relative '../day'
require_relative '../map'

class Day06 < Day
  class Problem
    attr_accessor :numbers, :operation

    def initialize(numbers = [], operation = :+)
      @numbers = numbers
      @operation = operation
    end

    def apply
      @operation == :+ ? 0 : 1
      @numbers.reduce { |acc, val| acc.send(@operation, val) }
    end
  end

  def do_part1(lines)
    parse_problems(lines).map(&:apply).sum
  end

  def do_part2(lines)
    parse_problems_the_cephalopod_way(lines).map(&:apply).sum
  end

  private

  def parse_problems(lines)
    number_rows = lines[..-2].map { |line| line.split.map(&:to_i) }
    ops = lines[-1].split.map(&:to_sym)
    problems = []
    ops.length.times do |i|
      problems << Problem.new(number_rows.map { |row| row[i] }, ops[i])
    end
    problems
  end

  def parse_problems_the_cephalopod_way(lines)
    m = Map.new(lines)
    problems = []
    problem = Problem.new
    (m.width - 1).downto(0).each do |col|
      n = 0
      op = nil
      m.height.times do |row|
        ch = m.at(row, col)
        if digit?(ch)
          n = n * 10 + ch.to_i
        elsif ['+', '*'].include?(ch)
          op = ch.to_sym
        end
      end
      next unless n != 0

      problem.numbers << n
      next if op.nil?

      problem.operation = op
      problems << problem
      problem = Problem.new
    end
    # ops are at left of column so we don't want to add the lastempty
    # problem to the problems array.
    problems
  end

  def digit?(ch)
    ch >= '0' && ch <= '9'
  end
end

if __FILE__ == $PROGRAM_NAME
  require_relative '../aoc'

  aoc(__FILE__)
end
